sys/include/net/gnrc/tx_sync.h

```diff
@@ -0,0 +1,140 @@
+ * @brief       This allows upper layers to wait for a transmission to complete
+ *              (or fail) and for passing up data about the transmission.
+ * @{
+ *
+ * @file
+ * @brief   Definitions for TX sync.
+ *
+ * @author  Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ */
+#ifndef NET_GNRC_TX_SYNC_H
+#define NET_GNRC_TX_SYNC_H
+
+#include <errno.h>
+#include <stdint.h>
+
+#include "mutex.h"
+#include "net/gnrc/pktbuf.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @brief   TX synchronization data */
+typedef struct {
+    mutex_t signal;     /**< Mutex used for synchronization */
+} gnrc_tx_sync_t;
+
+/**
+ * @brief   Helper to initialize a gnrc_tx_sync_t structure
+ */
+static inline gnrc_tx_sync_t gnrc_tx_sync_init(void)
+{
+    gnrc_tx_sync_t result = { .signal = MUTEX_INIT_LOCKED };
+    return result;
+}
+
+/**
+ * @brief   Build a TX sync snip
+ *
+ * @param[in,out]   tx_sync     TX sync structure the snip should hold
+ *
+ * @return  The TX sync snip holing @p tx_sync
+ * @retval  NULL    Allocation Failed
+ *
+ * @note    No need to initialize @p tx_sync, this function will do it for you.
+ */
+static inline gnrc_pktsnip_t * gnrc_tx_sync_build(gnrc_tx_sync_t *tx_sync)
+{
+    *tx_sync = gnrc_tx_sync_init();
+    gnrc_pktsnip_t *snip = gnrc_pktbuf_add(NULL, NULL, 0, GNRC_NETTYPE_TX_SYNC);
+    if (!snip) return NULL;
+    snip->data = tx_sync;
+    return snip;
+}
+
+/**
+ * @brief   Appends a newly allocated tx sync pktsnip to the end of the packet
+ *
+ * @param[in]       pkt     Packet to append TX sync pktsnip to
+ * @param[in,out]   tx_sync TX sync structure to initialize and append
+ *
+ * @retval  0       Success
+ * @retval  -ENOMEM Allocation failed
+ *
+ * @note    No need to initialize @p tx_sync, this function will do it for you.
+ */
+static inline int gnrc_tx_sync_append(gnrc_pktsnip_t *pkt,
+                                      gnrc_tx_sync_t *tx_sync)
+{
+    gnrc_pktsnip_t *snip = gnrc_tx_sync_build(tx_sync);
+    if (!snip) return -ENOMEM;
+    gnrc_pkt_append(pkt, snip);
+    return 0;
+}
+
+/**
+ * @brief   Split off the TX sync snip and return it
+ * @param[in,out]   pkt     Packet to split off the TX sync snip
+ * @return  The TX sync snip that no longer is part of @p pkt
+ * @retval  NULL            @p pkt contains no TX sync snip
+ */
+gnrc_pktsnip_t * gnrc_tx_sync_split(gnrc_pktsnip_t *pkt);
+
+/**
+ * @brief   Signal TX completion via the given tx sync packet snip
+ *
+ * @pre     Module gnrc_netttype_tx_sync is sued
+ * @pre     `pkt->type == GNRC_NETTYPE_TX_SYNC`
+ *
+ * @param   The tx sync packet snip of the packet that was transmitted
+ */
+static inline void gnrc_tx_complete(gnrc_pktsnip_t *pkt)
+{
+    assert(IS_USED(MODULE_GNRC_TX_SYNC) && (pkt->type == GNRC_NETTYPE_TX_SYNC));
+    /* Allow for multiple waiters by just unlocking the mutex until all
+     * blocked threads have resumed */
+    gnrc_tx_sync_t *sync = pkt->data;
+    do {
+        mutex_unlock(&sync->signal);
```

Comment:
Which API do you refer to by `signal`?

A condition variable is not really matching the use case here. For that you have a shared data structure which you can only access with a mutex locked. So you lock the mutex, check the state of the shared data structure if some condition is matched, and if not, you call `cond_wait()` with the mutex still locked as argument. `cond_wait()` will internally unlock the mutex will waiting, and lock it again before it returns. So just to comply with the API, we need an additional meaningless mutex, which adds complexity.

A semaphore is also not what we want here. A semaphore is a generalized mutex where the number of threads allowed to enter the critical core becomes an initialization parameter. We could block on it by initializing it with a capacity of zero. But unblocking all waiters again would require to call `sema_post()` in a loop until all waiters are unblocked. But without relying on implementation details, it is difficult to know when all waiters have been unblocked.

Anyway: I just added this "wake all waiters" feature as I saw that `gnrc_neterr` allows multiple threads to register (well, only one per snip). So if we want to base `gnrc_neterr` on this, I thought it might be required to still allow for multiple threads to wait for the error messages. However, in the RIOT code base only the very same threads registers for error reporting (but to every snip). So maybe it would be better to just allow a single thread to wait for completion. And if a use case pops up, this can still be implemented without touching the API. (And it would also allow using some different than a mutex for synchronization, if that turns out to be beneficial in such a use case.)

Reply:
I dropped the multi waiters feature in a fixup.

Reply:
> Which API do you refer to by `signal`?

I referred to your member name ;-)

Reply:
But otherwise okay.